test/functional/ui/incsubstitution_spec.lua

```diff
@@ -0,0 +1,1044 @@
+local clear, feed, insert = helpers.clear, helpers.feed, helpers.insert
+local execute, expect, eval = helpers.execute, helpers.expect, helpers.eval
+local neq, source, eq = helpers.neq, helpers.source, helpers.eq
+
+local default_text = [[
+  Inc substitution on
+  two lines
+]]
+
+local function common_setup(screen, incsub, text)
+  if screen then
+    execute("syntax on")
+    execute("set nohlsearch")
+    execute("hi IncSubstitute guifg=red guibg=yellow")
+    screen:attach()
+    screen:set_default_attr_ignore( {{bold=true, foreground=Screen.colors.Blue}} )
+    screen:set_default_attr_ids({
+      [1]  = {foreground = Screen.colors.Fuchsia},
+      [2]  = {foreground = Screen.colors.Brown, bold = true},
+      [3]  = {foreground = Screen.colors.SlateBlue},
+      [4]  = {bold = true, foreground = Screen.colors.SlateBlue},
+      [5]  = {foreground = Screen.colors.DarkCyan},
+      [6]  = {bold = true},
+      [7]  = {underline = true, bold = true, foreground = Screen.colors.SlateBlue},
+      [8]  = {foreground = Screen.colors.Slateblue, underline = true},
+      [9]  = {background = Screen.colors.Yellow},
+      [10] = {reverse = true},
+      [11] = {reverse = true, bold=true},
+      [12] = {foreground = Screen.colors.Red, background = Screen.colors.Yellow},
+      [13] = {bold = true, foreground = Screen.colors.SeaGreen},
+      [14] = {foreground = Screen.colors.White, background = Screen.colors.Red},
+    })
+  end
+
+  if incsub then
+    execute("set incsubstitute="..incsub)
+  else
+    execute("set incsubstitute=")
+  end
+
+  if text then
+    insert(text)
+  end
+end
+
+describe('IncSubstitution preserves', function()
+  before_each(clear)
+
+  it(':ls functionality', function()
+    local screen = Screen.new(30,10)
+    common_setup(screen, "split", "ABC")
+
+    execute("%s/AB/BA/")
+    execute("ls")
+
+    screen:expect([[
+      ~                             |
+      ~                             |
+      ~                             |
+      ~                             |
+      ~                             |
+      :ls                           |
+        1 %a + "[No Name]"          |
+                line 1              |
+      {13:Press ENTER or type command to}|
+      {13: continue}^                     |
+    ]])
+  end)
+
+  it('default substitution with various delimiters', function()
+    insert(default_text)
+    execute("set incsubstitute=")
+
+    local delims = { '/', '#', ';', '%', ',', '@', '!', ''}
+    for _,delim in pairs(delims) do
+      execute("%s"..delim.."lines"..delim.."LINES"..delim.."g")
+      expect([[
+        Inc substitution on
+        two LINES
+        ]])
+      execute("undo")
+    end
+  end)
+
+  it('the undolevels setting', function()
+    for _, case in pairs{"", "split", "nosplit"} do
+      clear()
+      execute("set undolevels=139")
+      execute("set incsubstitute=" .. case)
+      insert("as")
+      feed(":%s/as/glork/<enter>")
+      execute("put =&undolevels")
```

Comment:
*And* `curbufmeths.get_option('undolevels')`, will check both local and global values.